### PR TITLE
Update bucket conf to default AES256 encryption.

### PR DESCRIPTION
**Pull request by a-h**
## What did you implement:

Closes https://github.com/serverless/serverless/issues/5797

## How did you implement it:

Trivial change to the default bucket policy.

## How can we verify it:

Trivial change. Creating a simple Hello World function, and checking that the deployment bucket creating has default server-side encryption and versioning enabled via the AWS console would do it.

## Todos:

- [x] Write tests
- [x] Write documentation
- [x] Fix linting errors
- [x] Make sure code coverage hasn't dropped
- [x] Provide verification config / commands / resources
- [x] Enable "Allow edits from maintainers" for this PR
- [x] Update the messages below

***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO

### CODE REVIEW

**Comment by a-h** [comment]
The Node.js: 4.4 build failed, because it didn't recognise the spread operator, which I'd used to carry out Object.assign operations.

I could replace it with this?

```js
    mergeProperties.reduce((acc, val) => Object.assign(acc, val),
      this.serverless.service.provider.compiledCloudFormationTemplate
        .Resources.ServerlessDeploymentBucket.Properties);
```

File: lib/plugins/aws/package/lib/generateCoreTemplate.js [hunk 1]
```diff
@@ -54,27 +54,30 @@ module.exports = {
...
-        };
+      mergeProperties.push({
+        AccelerateConfiguration: {
+          AccelerationStatus: 'Suspended',
+        },
+      });
     }
 
+    Object.assign(this.serverless.service.provider.compiledCloudFormationTemplate
+      .Resources.ServerlessDeploymentBucket.Properties, ...mergeProperties);
```
Comment (pmuens) [inline lib/plugins/aws/package/lib/generateCoreTemplate.js]: `...` is not supported in Node.js 4 (which we still have to support 😞)

**Comment by exoego** [comment]
Enabling Encryption is welcome but what is a motivation to enable versioning on deployment bucket?

If I understand correctly, versioning skeeps history and increase sizes for deployment bucket.
IMO, I manage versions in my own git repo, so s3 versioning is not useful.

File: lib/plugins/aws/package/lib/generateCoreTemplate.js (hunk 1 above)
Comment (a-h) [reply lib/plugins/aws/package/lib/generateCoreTemplate.js]: Thanks, that makes sense, I'll sort that.

**Comment by a-h** [comment]
Versioning protects against accidental (or malicious deletion, especially when combined with MFA delete) deletion of the content and provides an audit history of all the code that was deployed.

Ensuring that versioning is enabled is a default rule within AWS Config and other AWS audit tools (e.g. Scout2). So the aim is to make the default Serverless config as production-ready as it can be / not stick out in audits and require later remediation.

Since S3 versioning keeps all old versions, then there are costs associated with the storage. This could be remediated by including a lifecycle policy to remove the old versions after a default period (e.g. 1 year?).

As to whether it should be the default, `versionFunctions` is set to true by default (ran out of space a few times now :)), so I thought that it might make sense to be default here.

There are no additional charges for enabling default S3 configuration, so that's a relatively easy win.

Happy drop versioning, or turn it into an option.

Maybe I should split versioning into another issue / PR and debate that there?

**Comment by darrenhgc** [comment]
> Happy to drop versioning, or turn it into an option. Maybe I should split versioning into another issue / PR and debate that there?

This. I believe this warrants more discussion.

**Comment by a-h** [comment]
Thanks for your enthusiastic comments, I feel very welcomed to the project!

Default encryption will be applied to existing buckets which were created by the Serverless Framework with this change. The encryption is transparent and uses Server-Side Encryption with Amazon S3-Managed Keys (SSE-S3), so there's no need to set a key, or permissions concerns as there would be might be with KMS-based encryption.

To test the effect on existing projects, I tested it on one of my projects which I'd previously deployed. The bucket had no encryption enabled and was using the default Serverless Framework configuration. After running the updated Serverless framework deployment, default encryption was then applied to the bucket, new objects added to the bucket (deployments) were encrypted, but objects that were already in the bucket (previous deployments) were not updated to be encrypted. I found no issues with the deployment or changes.

I've also tested it with a custom bucket configuration to make sure that custom configurations are not overwritten. I created an S3 bucket with versioning enabled, which uses a custom KMS key for encryption.

screenshot 2019-02-07 at 11 40 27

I then tested using a custom bucket instead of the default:

```yaml
provider:
  name: aws
  runtime: nodejs8.10
  deploymentBucket:
    name: xxxx-serverless-deployments
  stage: dev
  region: eu-west-2
```

I deployed it once using the current version of Serverless framework, then again using the version with this pull request applied. Afterwards, no changes were made to my custom bucket, as expected.

**Comment by ozbillwang** [comment]
@a-h 

Did you know your change made trouble and let these old codes failed? Are you going to fix them?

with exist example, 

https://github.com/serverless/examples/tree/master/aws-node-rest-api-with-dynamodb

https://github.com/serverless/examples/tree/master/aws-python-rest-api-with-dynamodb

Because of your PR, I am getting this error now

    An error occurred: ServerlessDeploymentBucket - API: s3:SetBucketEncryption Access Denied

I used to deploy before and it is fine.

There are so many useful examples, which I used to do show case, now I can't use them

**Comment by ozbillwang** [comment]
So I reviewed the code, and it set hard code to force use SSE as default. So there is no option I can avoid to use it, why?

**Comment by a-h** [comment]
Sorry for the issue it's causing you. It looks like an additional IAM permission is required by the role that you're using to execute the deployment, rather than the examples being broken.

The docs at https://www.serverless.com/framework/docs/providers/aws/guide/credentials/ specify the use of the Administrator role which does have the required permissions.

If you're using a custom role, then you need to add the following permissions to enable setting encryption on buckets:

```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Sid": "VisualEditor0",
            "Effect": "Allow",
            "Action": [ "s3:GetEncryptionConfiguration", "s3:PutEncryptionConfiguration" ],
            "Resource": "*"
        }
    ]
}
```
These are:

* s3:GetEncryptionConfiguration
* s3:PutEncryptionConfiguration

Here's the steps in the Web console:

Screenshot 2020-05-05 at 17 44 39

Screenshot 2020-05-05 at 17 45 04

Screenshot 2020-05-05 at 17 46 22

Screenshot 2020-05-05 at 17 50 47

If this is a problem for you, then you could use a custom bucket rather than letting Serverless Framework create it for you. I haven't tried this myself, but it's documented here: https://medium.com/@oieduardorabelo/reusing-s3-bucket-for-multiple-serverless-framework-projects-deploy-828e3a45f713

**Comment by ozbillwang** [comment]
@a-h 

Thanks for your kind to give details about this issue. But I have tried everything I can do.

I run `serverless deploy` with `AdministratorAccess` permission, I have deleted that user and re-create again with same `AdministratorAccess` policy. I have added that two permissions as well in that account, but still get same error.

**Comment by ozbillwang** [comment]
And again, @a-h . I still don't think that's good idea to hard code anything. 

If you want (from your own project experience or security compliance requirement), you active the s3 encryption configuraiton, but not hard code it. You should do that as option, that we can decide to set it or not.

**Comment by ozbillwang** [comment]
finally I got the root cause of this issue in my environment. 

Our company puts the **organization SCP (Service control policies)** to all sub OU accounts. Because of that, we have no permssion of `s3:PutEncryptionConfiguration` even I am in admin group/roles.

It will be hard to get the approval to release this block in short term (You know that's corp level's security review, it takes time), second it is a best practices of [AWS Landing zone](https://aws.amazon.com/about-aws/whats-new/2018/06/introducing-aws-landing-zone/) default SCP policy, in fact, security team would not approve this permission if I asked for.

We need rethink this PR and make this permission optional. That's the main problem we need fix here.

**Comment by dfang** [comment]
i'm using a member account in the root OU as master root in an organization, still got this issue.

**Comment by naeemhaq** [comment]
This solution is not working in a Landing Zone environment. I am surprised its still happening.

**Comment by ozbillwang** [comment]
@a-h 

Are you going to fix this issue, Which make your change optional, please

**Comment by a-h** [comment]
If your organisation's security policies are preventing you from creating an encrypted bucket and you're unable to update them, you can create a bucket outside of Serverless Framework and use that instead, there's a guide here https://medium.com/cloud-recipes/use-deploymentbucket-section-in-serverless-yml-179a4020b450

It's often a good idea to do that anyway, because you can quickly exceed the maximum number of S3 buckets in an account if you have lots of Serverless Framework projects.

I use a CloudFormation resource like this to create my single-bucket-per-account. This usually passes AWS configuration audits.

```yaml
  # Instead of using Serverless Framework to create a bucket, create your own so that we can control the policy more carefully.
  # This also reduces the number of buckets that are lying around.
  # The serverless.yaml can be configured to use this by adding a provider/deploymentBucket section.
  ServerlessDeploymentBucket:
    Type: AWS::S3::Bucket
    Properties:
      BucketEncryption:
        ServerSideEncryptionConfiguration:
          - ServerSideEncryptionByDefault:
              SSEAlgorithm: AES256
      LifecycleConfiguration: # Optional, can be used to reduce wasted storage if you don't need old versions to be retained.
        Rules:
          - AbortIncompleteMultipartUpload: # Abandon incomplete uploads after 7 days.
              DaysAfterInitiation: 7
            Status: Enabled
          - NoncurrentVersionExpirationInDays: 7 # Delete old versions of files after 7 days. Think about whether this meets your audit requirements.
            Status: Enabled
      PublicAccessBlockConfiguration: # Disallow public buckets.
        BlockPublicAcls: true
        BlockPublicPolicy: true
        IgnorePublicAcls: true
        RestrictPublicBuckets: true
      VersioningConfiguration: # Enable versioning.
        Status: Enabled
  ServerlessDeploymentBucketPolicy: # Policy to disable non HTTPS access, and enforce MFA delete.
    Type: AWS::S3::BucketPolicy
    Properties:
      Bucket: !Ref ServerlessDeploymentBucket
      PolicyDocument:
        Statement:
          - Sid: RequireTLS
            Effect: Deny
            Principal: "*"
            Action: "*"
            Resource:
              !Join ["", ["arn:aws:s3:::", !Ref ServerlessDeploymentBucket]]
            Condition:
              Bool:
                "aws:SecureTransport": false
          - Sid: RequireMFAForBucketDeletion
            Effect: Deny
            Principal: "*"
            Action:
              - s3:DeleteBucket
            Resource:
              !Join ["", ["arn:aws:s3:::", !Ref ServerlessDeploymentBucket]]
            Condition:
              "Null":
                "aws:MultiFactorAuthAge": true
``` 

@ozbillwang - feel free to submit a PR if you want to make the default encryption configuration optional, or indeed, a PR just to revert the change, and let the maintainers decide on the course of action.

**Comment by ozbillwang** [comment]
You still don't understand what is SCP  (service control policy). there is no chance for us to create a s3 bucket as you want to, because this permission has been taken from the Organization levels. 

The only way if let me take care is, frankly say, rollback your change.

If you need a s3 bucket with AES256 encryption, the way you show me just above is what you need to do.  

Please vote for me if anyone ask for a rollback of this change.

File: lib/plugins/aws/package/lib/core-cloudformation-template.json [hunk 2]
```diff
@@ -3,7 +3,18 @@
...
   "Resources": {
     "ServerlessDeploymentBucket": {
-      "Type" : "AWS::S3::Bucket"
+      "Type" : "AWS::S3::Bucket",
+      "Properties" : {
+        "BucketEncryption": {
+          "ServerSideEncryptionConfiguration": [
+            {
+              "ServerSideEncryptionByDefault": {
+                "SSEAlgorithm": "AES256"
```
Comment (ozbillwang) [inline lib/plugins/aws/package/lib/core-cloudformation-template.json:12]: That's the evil code.

**Pull request merged**